functions/src/services/GoogleSheet.ts

```diff
@@ -0,0 +1,358 @@
+
+enum IMajorDimensions {
+  Rows = 'ROWS',
+  Columns = 'COLUMNS',
+}
+
+enum IValueInputOption {
+  USER_ENTERED = 'USER_ENTERED',
+  RAW = 'RAW',
+}
+
+interface IValueRange {
+  range: string;
+  majorDimension: IMajorDimensions;
+  values: IRowValues[];
+}
+
+interface IGetSheetResponse {
+  headers: any[];
+  rows: any[];
+}
+
+abstract class SpreadSheet {
+  protected abstract connect(): any;
+  public abstract getSheet(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    limit?: number
+  ): Promise<IGetSheetResponse>;
+  public abstract appendRow(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    appendValues: any
+  ): Promise<any>;
+  public abstract getColumn(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    columnName: string
+  ): Promise<any>;
+  public abstract findRowWithColumnValue(
+    column: string,
+    rowValueToSearch: string | any,
+    tableToSearch: any
+  ): Promise<any>;
+}
+
+type IRowValues = string[] | number[];
+
+export default class GoogleSheets extends SpreadSheet {
+  protected connection: any;
+
+  protected async connect() {
+    const auth = await gapi.auth.getClient({
+      scopes: [GoogleScopes.SpreadSheets],
+    });
+    this.connection = await gapi.sheets({ version: 'v4', auth });
+    return this.connection;
+  }
+
+  /**
+   * Query for a specific google sheets within a spreadsheet
+   *
+   * @param sheet The sheet to query from google sheets api
+   * @param limit How many rows you want including the header titles
+   */
+  public async getSheet(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    limit?: number
+  ): Promise<IGetSheetResponse> {
+    await this.connect();
+    // console.log("Range query: ", sheet + this._computeRange(limit));
+    const targetSheet: any = await this.connection.spreadsheets.values.get({
+      spreadsheetId,
+      majorDimension: IMajorDimensions.Rows,
+      range: sheetName + this._computeRange(limit),
+    });
+
+    const { statusText, status, data, headers } = targetSheet;
+    if (statusText !== 'OK' || status !== 200) {
+      console.error('Error: Not able to get google sheet');
+      return null;
+    }
+    const { majorDimension, values }: IValueRange = data;
+    if (majorDimension !== IMajorDimensions.Rows || !values || !values.length) {
+      console.error('Error: Values are wrong format');
+      return null;
+    }
+
+    const headerKeys: any[] = values.shift();
+    return {
+      headers: headerKeys,
+      rows: this._convertRows(headerKeys, values),
+    };
+  }
+
+  public async getColumn(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    columnName: string
+  ) {
+    await this.connect();
+
+    // Retrieve first row of headers
+    const firstRow: any = await this.connection.spreadsheets.values.get({
+      spreadsheetId,
+      majorDimension: IMajorDimensions.Rows,
+      range: sheetName + '!A1:Z1',
```

Comment:
Z might not be the last column. I propose:
```suggestion
       range: sheetName + '!1:1',
```